import argparse
from configparser import ConfigParser, SectionProxy
from multiprocessing import cpu_count
from gui_frontend import main as gui_frontend_main
from install_resources import install_resources
import subprocess
import sys
from typing import Union


AVAILABLE_TOOLS = {
    "abricate",
    "isescan",
    "mobileog",
    "ectyper",
    "tncentral",
    "integron",
    "islandviewer",
    "plsdb",
    "parser",
}


def create_argument_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="GenomeProfiler",
        description="A tool to automate genome feature profiling and visualization.",
        epilog="""
Available Tools (use with --include):
  abricate       - Run abricate for resistance gene profiling.
  isescan        - Run ISEScan to identify insertion sequences.
  mobileog       - Run MobileOG-db analysis.
  ectyper        - Run ECTyper for serotyping.
  tncentral      - Run BLASTn search against TnCentral.
  integron       - Run Integron Finder.
  islandviewer   - Submit genome to IslandViewer for island prediction.
  plsdb          - Use PLSDB for plasmid similarity screening.
  phastest       - Run local PHASTEST search for phage regions.
  parser         - Parse tool outputs for visualization (enabled by default in GUI).

Examples:
  python cli_main.py NZ_CP000000
  python cli_main.py NZ_CP000000 --include-tools abricate parser --timestamped-output
  python cli_main.py --help
        """,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )

    parser.add_argument(
        "accessions",
        nargs="*",
        help="NCBI accession(s) to process (optional if using --fasta or --genbank)",
    )
    parser.add_argument(
        "--setup",
        action="store_true",
        help="Download resources required by tools, such as databases. NOTE: Run using this once before using other options.",
    )
    parser.add_argument(
        "-g",
        "--gui",
        action="store_true",
        help="Run GenomeProfiler in graphical mode. NOTE: This will ignore all other arguments.",
    )
    parser.add_argument(
        "-t",
        "--include",
        nargs="*",
        default=[],
        help="Run these tools on the accessions.",
    )
    parser.add_argument(
        "-a",
        "--all-tools",
        action="store_true",
        help="Run all available tools on the accessions.",
    )
    parser.add_argument(
        "--exclude",
        nargs="*",
        default=[],
        help="Do not run these tools on the accessions. NOTE: If a tool is included and excluded, it will not be included.",
    )
    parser.add_argument(
        "--fasta",
        help="Optional local FASTA file to use instead of fetching from NCBI",
    )
    parser.add_argument(
        "--genbank",
        help="Optional local GenBank file to use instead of fetching from NCBI",
    )
    parser.add_argument(
        "-j",
        "--workers",
        type=int,
        help=f"Number of concurrent workers (default: {int(cpu_count() / 2)}, half of CPU count)",
    )
    parser.add_argument(
        "-o",
        "--output-dir",
        help="Base directory to store results (overrides: output_base in config; default: output_GenomeProfiler)",
    )
    parser.add_argument(
        "-r",
        "--resource-dir",
        default="GenomeProfiler/resource",
        help="Path to resource directory generated by --setup (default: GenomeProfiler/resources)",
    )
    parser.add_argument(
        "-c",
        "--config",
        help="Path to configuration file",
    )
    parser.add_argument(
        "-n",
        "--no-interact",
        action="store_true",
        help="Terminate instead of prompting for input. NOTE: Useful for scripts.",
    )
    parser.add_argument(
        "--sleep-interval",
        type=int,
        help="(in seconds) Delay between status polls to IslandViewer.",
    )

    return parser


def create_resource_paths(resource_base_path: str) -> dict:
    return {
        "plsdb_meta_dir": f"{resource_base_path}/plsdb-meta-2024_05_31_v2",
        "plsdb_sketch_path": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/plsdb_sketch.msh",
        "assembly_csv": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/assembly.csv",
        "nuccore_csv": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/nuccore.csv",
        "plasmidfinder.csv": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/plasmidfinder.csv",
        "typing.csv": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/typing.csv",
        "taxonomy.csv": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/taxonomy.csv",
        "biosample.csv": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/biosample.csv",
        "changes.tsv": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/changes.tsv",
        "amr.tsv": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/amr.tsv",
        "plsdb_mashdb_sim.tsv": f"{resource_base_path}/plsdb-meta-2024_05_31_v2/plsdb_mashdb_sim.tsv",
        "mobileog_db_faa": f"{resource_base_path}/mobileOG-db-beatrix-1.6/mobileOG-db_beatrix-1.6.All.faa",
        "mobileog_db_csv": f"{resource_base_path}/mobileOG-db-beatrix-1.6/mobileOG-db-beatrix-1.6-All.csv",
        "tncentral_db": f"{resource_base_path}/tncentral-db",
        "tncentral_fasta": f"{resource_base_path}/tncentral-db/tncentral_integrall_isfinder.fa",
        "phastest_db": f"{resource_base_path}/phastest/prophage_virus.db",
    }


def create_config_overrides(args: argparse.Namespace) -> dict:
    overrides = create_resource_paths(args.resource_dir)

    if args.workers:
        overrides["max_workers"] = str(args.workers)
        overrides["ectyper_cores"] = str(args.workers)

    if args.output_dir:
        overrides["output_base"] = args.output_dir

    if args.sleep_interval:
        overrides["sleep_interval"] = str(args.sleep_interval)

    return overrides


def load_and_resolve_config_file(
    path: Union[None, str],
    section: str,
    overrides: dict,
) -> Union[None, SectionProxy]:
    DEFAULT_ENTRIES = {
        "islandviewer_api_submit": "https://www.pathogenomics.sfu.ca/islandviewer/http_api/rest/submit/",
        "islandviewer_api_status": "https://www.pathogenomics.sfu.ca/islandviewer/rest/job/{token}/",
        "islandviewer_api_download": "https://www.pathogenomics.sfu.ca/islandviewer/rest/job/{token}/download/tab/",
        "abricate_path": "abricate",
        "integron_path": "integron_finder",
        "isescan_path": "isescan.py",
        "ectyper_path": "ectyper",
        "prodigal_path": "prodigal",
        "diamond_path": "diamond",
        "output_base": "output_GenomeProfiler",
        "max_workers": str(int(cpu_count() / 2)),
        "sleep_interval": "300",
        "ectyper_cores": str(int(cpu_count() / 2)),
    }

    parser = ConfigParser()

    # From config file
    if path:
        if not parser.read(path) or not parser.has_section(section):
            raise EnvironmentError(f"Configuration file not found at '{path}'")
    else:
        parser.read_dict({"genome_profiler": {}})

    # From defaults not in config file
    for key in DEFAULT_ENTRIES:
        if not parser.has_option(section, key):
            parser.set(section, key, DEFAULT_ENTRIES[key])

    # From overrides
    for key in overrides:
        parser.set(section, key, overrides[key])

    return parser


def validate_environment(config: SectionProxy):
    checks = [
        ("abricate", [config["abricate_path"], "--version"]),
        ("integron_finder", [config["integron_path"], "--version"]),
        ("isescan", [config["isescan_path"], "--version"]),
        ("ectyper", [config["ectyper_path"], "--version"]),
        ("diamond", [config["diamond_path"], "version"]),
        ("prodigal", [config["prodigal_path"], "-v"]),
    ]

    missing = []
    for name, cmd in checks:
        try:
            subprocess.run(
                cmd, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
            )
        except (subprocess.CalledProcessError, FileNotFoundError):
            missing.append(name)

    if missing:
        raise EnvironmentError(f"Missing tools: {', '.join(missing)}")


def resolve_config_and_args() -> SectionProxy:
    """
    Priority of options (descending):
    1. CLI arguments
    2. Config
    3. Defaults
    """

    cli_parser = create_argument_parser()

    args = cli_parser.parse_args()

    if len(sys.argv) == 1:
        cli_parser.print_help()
        sys.exit(0)

    if args.gui:
        gui_frontend_main()
        sys.exit(0)

    config_overrides = create_config_overrides(args)

    config = load_and_resolve_config_file(
        args.config,
        "genome_profiler",
        config_overrides,
    )

    if args.setup:
        install_resources()
        sys.exit(0)

    return config


def main():
    config = resolve_config_and_args()
    # validate_environment(config)

    # FIXME debug
    print(vars(config)["_sections"])


if __name__ == "__main__":
    main()
